lib/Numeric.hs

```diff
@@ -1,14 +1,113 @@
+  showBin,
+  showHex,
+  showOct,
+  showEFloat,--
+  showFFloat,
+  showGFloat,
+  showFloat,
+  floatToDigits,
   readSigned,
   readInt,
   readBin,
   readDec,
   readOct,
   readHex,
-  readIntegral,
+  readFloat,
+  lexDigits,
+  fromRat,
   ) where
-import qualified Prelude()              -- do not import Prelude
+
+import Primitives
+import Data.Integer_Type
+import Data.Integer.Internal
+import Data.Ratio_Type
+import Data.Ratio
+import Numeric.FormatFloat
 import Numeric.Read
 import Numeric.Show
+import Text.ParserCombinators.ReadP (ReadP, pfail, readP_to_S)
+import Text.Read.Internal (ReadS, lexDigits)
+import qualified Text.Read.Lex as L
+
+-- | Reads an /unsigned/ 'RealFrac' value,
+-- expressed in decimal scientific notation.
+readFloat :: RealFrac a => ReadS a
+readFloat = readP_to_S readFloatP
+
+readFloatP :: RealFrac a => ReadP a
+readFloatP =
+  do tok <- L.lex
+     case tok of
+       L.Number n -> return $ fromRational $ L.numberToRational n
+       _          -> pfail
+
+-- | Converts a 'Rational' value into any type in class 'RealFloat'.
+fromRat :: (RealFloat a) => Rational -> a
+
+-- Deal with special cases first, delegating the real work to fromRat'
+fromRat (n :% 0) | n > 0     =   1/0       -- +Infinity
+                 | n < 0     = -(1/0)      -- -Infinity
+                 | otherwise =   0/0       -- NaN
+
+fromRat (n :% d) | n > 0     = fromRat' (n :% d)
+                 | n < 0     = - fromRat' ((-n) :% d)
+                 | otherwise = encodeFloat 0 0             -- Zero
+
+-- Conversion process:
+-- Scale the rational number by the RealFloat base until
+-- it lies in the range of the mantissa (as used by decodeFloat/encodeFloat).
+-- Then round the rational to an Integer and encode it with the exponent
+-- that we got from the scaling.
+-- To speed up the scaling process we compute the log2 of the number to get
+-- a first guess of the exponent.
+
+fromRat' :: (RealFloat a) => Rational -> a
+-- Invariant: argument is strictly positive
+fromRat' x = r
+  where b = floatRadix r
+        p = floatDigits r
+        (minExp0, _) = floatRange r
+        minExp = minExp0 - p            -- the real minimum exponent
+        xMax   = toRational (b ^ p)
+        ln     = integerLogBase b (numerator x)
+        ld     = integerLogBase b (denominator x)
+        p0     = (ln - ld - p) `max` minExp
+        -- if x = n/d and ln = integerLogBase b n, ld = integerLogBase b d,
+        -- then b^(ln-ld-1) < x < b^(ln-ld+1)
+        f = if p0 < 0 then 1 :% (b ^ (-p0)) else (b ^ p0) :% 1
+        x0 = x / f
+        -- if ln - ld >= minExp0, then b^(p-1) < x0 < b^(p+1), so there's at most
+        -- one scaling step needed, otherwise, x0 < b^p and no scaling is needed
+        (x', p') = if x0 >= xMax then (x0 / toRational b, p0+1) else (x0, p0)
+        r = encodeFloat (round x') p'
+
+wordLog2 :: Word -> Int
+wordLog2 w = _wordSize - 1 - primWordClz w
+
+integerLog2 :: Integer -> Int
+integerLog2 (I sign ds) =
```

Comment:
This won't work if you compile to use GMP.  The internal representation of Integer should never be used.
Perhaps the integerLog stuff should move to some other module.

I also think that the fromRat code is much more complex than necessary (I don't mind saying that, since I'm the original author of the code, almost 35 years ago).  I think a variation of `fromRational` from `Fractional Double` would be better.

Reply:
> This won't work if you compile to use GMP.

Right, of course. There should probably be a CI job using GMP.

> I think a variation of `fromRational` from `Fractional Double` would be better.

Then I have some questions on [that code](https://github.com/augustss/MicroHs/blob/e72d01f5be926255fe25320f2fec83bdc90ce919/lib/Data/Double.hs#L49-L64):
- Why 1023 instead of 1024 (which is the maximum exponent)?
- Why require `d < maxVal` instead of `d <= maxVal`?
- How was 256 chosen as shift?

I don't think anyone really uses `fromRat` (most people probably don't even know that it exists), so it's not important for it to be very efficient. I'm happy to go with your suggestion, if that's what you prefer.

Reply:
Yes, a CI job for GMP would be great.  It really only has to be run on one platform.

Answers: all the answers are that it was just accidental.  It doesn't really matter exactly what the constants are, as long as they scale numbers that are too big.  It should be possible to figure out some better set of constants from the `RealFloat` information.  In which case `Double` and `Float` could share a `fromRational` implementation.